### PR TITLE
Migrate most of  debugging/metadata and loaders directories to null safety

### PR DESCRIPTION
**Pull request by annagrin**
Migrate most of ` debugging/metadata` and `loaders` directories to null safety.

### CODE REVIEW

File: dwds/lib/src/debugging/metadata/module_metadata.dart [hunk 1]
```diff
@@ -62,25 +60,25 @@ class ModuleMetadataVersion {
 /// See: https://goto.google.com/dart-web-debugger-metadata
 class LibraryMetadata {
   /// Library name as defined in pubspec.yaml
-  final String name;
+  late final String name;
```
Comment (annagrin) [inline dwds/lib/src/debugging/metadata/module_metadata.dart:63]: remove 'late' from those?

Comment (annagrin) [reply dwds/lib/src/debugging/metadata/module_metadata.dart]: done!

File: dwds/lib/src/loaders/build_runner_require.dart [hunk 2]
```diff
@@ -51,9 +52,18 @@ class BuildRunnerRequireStrategyProvider {
       throw StateError('Could not read digests at path: $digestsPath');
     }
     final body = await response.readAsString();
+    final digests = json.decode(body) as Map<String, dynamic>;
+
+    for (final key in digests.keys) {
```
Comment (grouma) [inline dwds/lib/src/loaders/build_runner_require.dart:57]: Why do we need this additional bit of logic?

File: dwds/lib/src/loaders/build_runner_require.dart [hunk 3]
```diff
@@ -63,9 +73,17 @@ class BuildRunnerRequireStrategyProvider {
...
-          stripTopLevelDirectory(key), value))[relativizePath(serverPath)];
+      MetadataProvider metadataProvider, String serverPath) async {
+    final modulePathToModule = await metadataProvider.modulePathToModule;
+    final relativePath = relativizePath(serverPath);
+    for (var e in modulePathToModule.entries) {
+      if (stripTopLevelDirectory(e.key) == relativePath) {
+        return e.value;
+      }
+    }
+    _logger.warning('No module for server path $serverPath');
```
Comment (grouma) [inline dwds/lib/src/loaders/build_runner_require.dart:84]: Should we be throwing or returning a null value here instead?

File: dwds/lib/src/loaders/frontend_server_require.dart [hunk 4]
```diff
@@ -56,7 +55,12 @@ class FrontendServerRequireStrategyProvider {
   Future<String> _moduleForServerPath(
       MetadataProvider metadataProvider, String serverPath) async {
     final modulePathToModule = await metadataProvider.modulePathToModule;
-    return modulePathToModule[_removeBasePath(serverPath)];
+    final relativeServerPath = _removeBasePath(serverPath);
+    if (!modulePathToModule.containsKey(relativeServerPath)) {
+      _logger.warning('No module found for server path: $serverPath.');
+      return '';
```
Comment (grouma) [inline dwds/lib/src/loaders/frontend_server_require.dart:61]: Same here. Should we be throwing or returning null instead?

**Review by grouma — APPROVED** [review]

File: dwds/lib/src/loaders/build_runner_require.dart (hunk 2 above)
Comment (annagrin) [reply dwds/lib/src/loaders/build_runner_require.dart:57]: The code below in its previous version would create weird entries in the returned map if a digest key is not present in modules:  { "null": value }, and we did not report them. I removed those (see code below) and added a warning just so we notice when unknown keys are present in digests so we investigate those cases if and when it happens.

File: dwds/lib/src/loaders/frontend_server_require.dart (hunk 4 above)
Comment (annagrin) [reply dwds/lib/src/loaders/frontend_server_require.dart]: Added throwing instead, thanks!

Comment (elliette) [reply dwds/lib/src/loaders/frontend_server_require.dart]: One concern about throwing - in the past we have crashed Flutter tools by throwing (see https://github.com/dart-lang/webdev/pull/1642). We should make sure that if this does throw it doesn't crash the running Flutter app.

File: dwds/lib/src/loaders/build_runner_require.dart (hunk 3 above)
Comment (annagrin) [reply dwds/lib/src/loaders/build_runner_require.dart]: Oops, throwing breaks our flow (some non-module paths are passed to those API), changed to allowing null instead.

File: dwds/lib/src/loaders/frontend_server_require.dart (hunk 4 above)
Comment (annagrin) [reply dwds/lib/src/loaders/frontend_server_require.dart]: Oops, throwing breaks our flow (some non-module paths are passed to those API), changed to allowing null instead.

For the general case of dwds crashing we should add a catch-all statement in flutter tools, as wells as a warning and a test  as well.

**Pull request merged**